codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2DataFrame.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    /**
+     * Construct a new data message.
+     *
+     * @param content non-{@code null} payload
+     * @param endStream whether this data should terminate the stream
+     * @param padding additional bytes that should be added to obscure the true content size
+     */
+    public DefaultHttp2DataFrame(ByteBuf content, boolean endStream, int padding) {
+        this.content = checkNotNull(content, "content");
+        this.endStream = endStream;
+        if (padding < 0 || padding > Http2CodecUtil.MAX_UNSIGNED_BYTE) {
+            throw new IllegalArgumentException("padding must be non-negative and less than 256");
+        }
+        this.padding = padding;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame setStream(Object stream) {
+      super.setStream(stream);
+      return this;
+    }
+
+    @Override
+    public boolean endStream() {
+        return endStream;
+    }
+
+    @Override
+    public int padding() {
+        return padding;
+    }
+
+    @Override
+    public ByteBuf content() {
+        if (content.refCnt() <= 0) {
+            throw new IllegalReferenceCountException(content.refCnt());
+        }
+        return content;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame copy() {
+        return new DefaultHttp2DataFrame(content().copy(), endStream, padding);
+    }
+
+    @Override
+    public DefaultHttp2DataFrame duplicate() {
+        return new DefaultHttp2DataFrame(content().duplicate(), endStream, padding);
+    }
+
+    @Override
+    public int refCnt() {
+        return content.refCnt();
+    }
+
+    @Override
+    public boolean release() {
+        return content.release();
+    }
+
+    @Override
+    public boolean release(int decrement) {
+        return content.release(decrement);
+    }
+
+    @Override
+    public DefaultHttp2DataFrame retain() {
+        content.retain();
+        return this;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame retain(int increment) {
+        content.retain(increment);
+        return this;
+    }
+
+    @Override
+    public String toString() {
+        return "DefaultHttp2DataFrame(stream=" + stream() + ",content=" + content
+            + ",endStream=" + endStream + ",padding=" + padding + ")";
+    }
+
+    @Override
+    public DefaultHttp2DataFrame touch() {
+        content.touch();
+        return this;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame touch(Object hint) {
+        content.touch(hint);
+        return this;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof DefaultHttp2DataFrame)) {
+            return false;
+        }
+        DefaultHttp2DataFrame other = (DefaultHttp2DataFrame) o;
+        return super.equals(other) && content.equals(other.content())
+            && endStream == other.endStream && padding == other.padding;
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = super.hashCode();
+        hash = hash * 31 + content.hashCode();
```

Comment:
should we consider the case if the content is released? maybe cache the hash code the first time it is called?